lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -545,84 +546,192 @@ def as_dict_with_promql_labels(self):
+    # - alert groups (plural): the list of groups[] (a list, i.e. no "groups:" key) - it is a list
+    #   of dictionaries that have the "name" and "rules" keys.
+    # - alert group (singular): a single dictionary that has the "name" and "rules" keys.
+    # - alert rules (plural): all the alerts in a given alert group - a list of dictionaries with
+    #   the "alert" and "expr" keys.
+    # - alert rule (singular): a single dictionary that has the "alert" and "expr" keys.
 
-            # insert juju topology filters into a prometheus alert rule
-            rule["expr"] = topology.render(rule["expr"])
-            return rule
+    def __init__(self, topology: JujuTopology):
+        self.topology = topology
+        self.alert_groups = []  # type: List[dict]
 
+    @staticmethod
+    def _is_official_format(rules_dict: dict) -> bool:
+        # TODO validate e.g. against schema using pydantic
+        return "groups" in rules_dict
 
-def load_alert_rules_from_dir(
-    dir_path: Union[str, Path], topology: JujuTopology, *, recursive: bool = False
-) -> List[dict]:
-    """Load alert rules from rule files.
+    @staticmethod
+    def _is_custom_format(rules_dict: dict) -> bool:
+        # one alert rule per file
+        # TODO validate e.g. against schema using pydantic
+        return set(rules_dict) >= {"alert", "expr"}
 
-    All rules from files for the same directory are loaded into a single
-    group. The generated name of this group includes juju topology.
-    By default, only the top directory is scanned; for nested scanning, pass `recursive=True`.
+    def _from_file(self, root_path: Path, file_path: Path) -> List[dict]:
+        """Read a rules file from path, injecting juju topology.
 
-    Args:
-        dir_path: directory containing *.rule files (alert rules without groups).
-        topology: a `JujuTopology` instance.
-        recursive: flag indicating whether to scan for rule files recursively.
+        Args:
+            root_path: full path to the root rules folder (used only for generating group name)
+            file_path: full path to a *.rule file.
 
-    Returns:
-        a list of prometheus alert rule groups.
-    """
-    alerts = defaultdict(list)
+        Returns:
+            A list of dictionaries representing the rules file, if file is valid (the structure is
+            formed by `yaml.safe_load` of the file); an empty list otherwise.
+        """
+        with file_path.open() as rf:
+            # Load a list of rules from file then add labels and filters
+            try:
+                rule_file = yaml.safe_load(rf)
+
+            except Exception as e:
+                logger.error("Failed to read alert rules from %s: %s", file_path.name, e)
+                return []
+
+            # TODO validate input
+            if AlertRules._is_official_format(rule_file):
+                alert_groups = rule_file["groups"]
+            elif AlertRules._is_custom_format(rule_file):
```

Comment:
official and custom seems like unnecessary signaling, especially as we want to promote the use of the "custom format".

Reply:
What do you mean? Are you proposing to rename them e.g. to "prom format" and "charm format"?
I was not aware we want to promote the custom format.

Reply:
If by the custom format you mean single rule, then yes. But skip it for now. I'm ok with this.